tools/src/main/java/org/terracotta/utilities/io/Files.java

```diff
@@ -796,6 +801,101 @@ private static Path copyInternal(Path source, Path target, CopyOption[] options)
+   * other exceptions are rethrown.
+   * <p>
+   * The caller is responsible for closing the input {@code stream}.
+   * <h3>Notes</h3>
+   * When {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} is specified, this method first calls
+   * {@link #delete(Path) delete(target)} before attempting to open an output stream on {@code target}.
+   * Because of this, there is a race window between deleting the target and opening target for output during
+   * which another process may create a file or directory named as {@code target}.  Handling of other process
+   * interference and pending deletion is handled during delete but no interference detection is performed while
+   * opening target for output -- interference at this time may result in an {@link AccessDeniedException}
+   * or another exception being thrown.
+   * @param stream the {@code InputStream} to be copied to {@code target}
+   * @param target the path of the file into which the {@code InputStream} is written
+   * @param options {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} can be used to
+   *    delete {@code target} if it exists; other options are not supported
+   * @return the number of bytes copied
+   * @throws DirectoryNotEmptyException if {@code target} is a non-empty directory and
+   *          {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} is specified
+   * @throws FileAlreadyExistsException if {@code target} exists and
+   *          {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} is not specified
+   * @throws IOException if the copy operation fails
+   * @throws NullPointerException if {@code stream} or {@code target} is null
+   * @throws UnsupportedOperationException if {@code options} contains an unsupported value
+   * @throws SecurityException if permission to access the target file/directory is not sufficient
+   * @see java.nio.file.Files#copy(InputStream, Path, CopyOption...)
+   */
+  public static long copy(InputStream stream, Path target, CopyOption... options) throws IOException {
+    Objects.requireNonNull(stream, "stream must be non-null");
+    Objects.requireNonNull(target, "target must be non-null");
+
+    Set<CopyOption> copyOptions = new LinkedHashSet<>(Arrays.asList(options));
+    if (!EnumSet.of(REPLACE_EXISTING).containsAll(copyOptions)) {
+      copyOptions.remove(REPLACE_EXISTING);
+      throw new UnsupportedOperationException("Unsupported option(s) specified - " + copyOptions);
+    }
+    boolean replace = copyOptions.contains(REPLACE_EXISTING);
+
+    Path resolvedTarget = target.toAbsolutePath();
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("copy(InputStream, {}, {})", resolvedTarget, Arrays.toString(options));
+    }
+
+    SecurityManager securityManager = System.getSecurityManager();
+    if (securityManager != null) {
+      securityManager.checkPermission(new FilePermission(resolvedTarget.toString(), "read,write" + (replace ? ",delete" : "")));
```

Comment:
I hate that the constants for these are private... any chance of having a static factory and an enum for this privately in this class.

Reply:
Believe it or not, the actions (`read`, `write`, etc.) are _documented_ in [java.io.FilePermission](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FilePermission.html).  The Java I/O classes are full of constants for which there _should_ have been public fields (or enumerations).  I like your idea but a separate PR seems the right way to go there.